openedx/core/djangoapps/content/course_structures/api/v0/api.py

```diff
@@ -61,14 +63,39 @@ def course_structure(course_key):
     Raises:
         CourseStructureNotAvailableError, CourseNotFoundError
     """
-    course = _retrieve_course(course_key)
-    try:
-        requested_course_structure = models.CourseStructure.objects.get(course_id=course.id)
-        return serializers.CourseStructureSerializer(requested_course_structure.structure).data
-    except models.CourseStructure.DoesNotExist:
-        # If we don't have data stored, generate it and return an error.
-        tasks.update_course_structure.delay(unicode(course_key))
-        raise CourseStructureNotAvailableError
+    modified_timestamp = models.CourseStructure.objects.filter(course_id=course_key).values('modified')
+
+    if modified_timestamp.exists():
+        cache_key = 'openedx.content.course_structures.api.v0.api.course_structure.{}.{}.{}'.format(
+            course_key, modified_timestamp[0]['modified'], '_'.join(block_types or [])
+        )
+        data = cache.get(cache_key)  # pylint: disable=maybe-no-member
+        if data:
+            return data
```

Comment:
1. You can set the cache timeout to forever (`None`), since it should never change for that course+timestamp.
2. Please put an explicit check for `if data is not None:` -- it doesn't really matter in this case, but there are cases where an empty dict or list is ok, and people can miss that more easily if they copy-paste this idiom.